### ISSUE TITLE
Getting confidence score and label when using InvokeEndpoint with XgBoost model

### ISSUE THREAD

**Issue by lujea** [aws/aws-sdk-java#1584]
To provide some context I am trying to build a model for a document classification problem using text documents. Each document receive one label and right now there is only two labels possible.
I am using the SageMaker Java client and it works fine to train and deploy a model using XgBoost algorithm. 
However, what I am not able to achieve is to get a confidence score together with the model prediction. 

I have tried training a model with different hyperparameter set however I cannot figure out how to get the confidence score with the prediction that is coming back. Here is the hyperparameters that I have tried:
num_round = 20
objective = binary:logistic
eval_metric = error
When I call InvoreEndpoint with that model, I get back a number between 0 and 1 but the actual class label.

I have also tried to set the hyperparameters to:
num_round = 20
objective = multi:softprob
num_class = 2
eval_metric = error
But I still don't get a confidence score, in both case the invokoEndpoint method returns a string with a number.
Can you please help to figure out how to get the confidence score with the label ?

**Comment by dagnir** [comment]
Hi @lujea, I couldn't tell from your question if you need usage guidance or if you're experiencing a problem using the SDK. Since we have limited bandwidth to respond to SDK issues on GitHub, we use this issues page only for bugs/feature requests in the SDK. If you're looking for guidance on using SageMaker, please open a question on [StackOverflow](http://stackoverflow.com/questions/tagged/aws-java-sdk) and tag it with `aws-java-sdk`.

If this is a bug, please provide details on your SDK version as well as actual and expected behavior.

**Comment by lujea** [comment]
Hi Dongie,

Thanks for your reply. I think this is bug as I am expecting the 
InvokeEndpoint method to return with a label and the confidence score 
related to that label. According to this link 
https://docs.aws.amazon.com/sagemaker/latest/dg/tf-example1-invoke.html 
this is the output of an TensorFlow model:

|{u'result': {u'classifications': [{u'classes': [{u'label': u'0', 
u'score': 0.009605729021131992}, {u'label': u'1', u'score': 
0.9699361324310303}, {u'label': u'2', u'score': 0.02045818418264389}]}]}}|


I would expect the same king of output for Xgboost. Currently I only get 
a string that contains a number between 0 and 1.

Ludovic

**Comment by lujea** [comment]
Will come back to you as soon as I get the logs.

Ludovic

**Comment by lujea** [comment]
@dagnir I have sent you a file with the traces, please let me know if you have it ?

**Comment by lujea** [comment]
@dagnir if there is some specific part that you want to see I can also paste them here for you.

**Comment by dagnir** [comment]
Just the response from the server should be fine. I'm curious to see if it's an unmarshalling issue.

**Comment by lujea** [comment]
@dagnir do you need more traces or is that enough ?

**Comment by dagnir** [comment]
@lujea What i'm looking for should actually be somewhere after the stuff you posted. It will look something like this (first part is the request, the second part is the response):

```
2018-05-09 15:14:36,296 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "POST /endpoints/my-endpoint/invocations HTTP/1.1[\r][\n]"
2018-05-09 15:14:36,296 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "Host: runtime.sagemaker.us-west-2.amazonaws.com[\r][\n]"
2018-05-09 15:14:36,297 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "Authorization: <AUTH>
2018-05-09 15:14:36,297 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "X-Amz-Date: 20180509T221435Z[\r][\n]"
2018-05-09 15:14:36,297 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "Accept: application/json[\r][\n]"
2018-05-09 15:14:36,297 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "User-Agent: aws-internal/3[\r][\n]"
2018-05-09 15:14:36,297 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "amz-sdk-invocation-id: 3ec4be64-035b-94f9-8641-8b4168301480[\r][\n]"
2018-05-09 15:14:36,297 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "amz-sdk-retry: 0/0/500[\r][\n]"
2018-05-09 15:14:36,297 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "Content-Type: [\r][\n]"
2018-05-09 15:14:36,297 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "Transfer-Encoding: chunked[\r][\n]"
2018-05-09 15:14:36,297 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "Connection: Keep-Alive[\r][\n]"
2018-05-09 15:14:36,297 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "[\r][\n]"
2018-05-09 15:14:36,298 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "8[\r][\n]"
2018-05-09 15:14:36,298 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "[0x0][0x0][0x0][0x0][0x0][0x0][0x0][0x0][\r][\n]"
2018-05-09 15:14:36,298 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "0[\r][\n]"
2018-05-09 15:14:36,298 [main] DEBUG org.apache.http.wire -  http-outgoing-0 >> "[\r][\n]"
2018-05-09 15:14:36,349 [main] DEBUG org.apache.http.wire -  http-outgoing-0 << "HTTP/1.1 400 Bad Request[\r][\n]"
2018-05-09 15:14:36,350 [main] DEBUG org.apache.http.wire -  http-outgoing-0 << "Date: Wed, 09 May 2018 22:14:36 GMT[\r][\n]"
2018-05-09 15:14:36,350 [main] DEBUG org.apache.http.wire -  http-outgoing-0 << "Content-Type: application/json[\r][\n]"
2018-05-09 15:14:36,350 [main] DEBUG org.apache.http.wire -  http-outgoing-0 << "Content-Length: 101[\r][\n]"
2018-05-09 15:14:36,350 [main] DEBUG org.apache.http.wire -  http-outgoing-0 << "Connection: keep-alive[\r][\n]"
2018-05-09 15:14:36,350 [main] DEBUG org.apache.http.wire -  http-outgoing-0 << "x-amzn-RequestId: 4573d7db-30b3-4ecd-8e81-34162ba7f4eb[\r][\n]"
2018-05-09 15:14:36,350 [main] DEBUG org.apache.http.wire -  http-outgoing-0 << "x-amzn-ErrorType: ValidationError:http://internal.amazon.com/coral/com.amazonaws.sagemaker.runtime/[\r][\n]"
2018-05-09 15:14:36,351 [main] DEBUG org.apache.http.wire -  http-outgoing-0 << "[\r][\n]"
```

**Comment by lujea** [comment]
@dagnir I have look at the traces but I can't find any trace looking that the one you posted. Am I missing something?

**Comment by lujea** [comment]
I see this trace:
{sagemaker, us-east-1} was not found in region metadata, trying to construct an endpoint using the standard pattern for this region: 'sagemaker.us-east-1.amazonaws.com'.
Could that be that something is wrong with the endpoint ?

**Comment by dagnir** [comment]
@lujea Thanks but the traces are still incomplete unfortunately. It does not include the response from SageMaker. Are you able to provide the full traces? It might be easier than trying to get just the request/response. Note you can attach file directly to your comment.

**Comment by lujea** [comment]
@dagnir thanks I thought I pasted enough. I have uploaded the file here:
https://drive.google.com/file/d/1yfE43GhzxVJUH8mmfZySQePuCeCTqJcZ/view?usp=sharing

Let me know if you can access it.

**Comment by dagnir** [comment]
Okay I took a look at the logs and it looks like SageMaker is really just returning the score as seen in this response

```
2018-05-10 09:31:35,336 DEBUG [pool-7-thread-1] o.a.http.wire [Wire.java:72] http-outgoing-37 << "HTTP/1.1 200 OK[\r][\n]"
2018-05-10 09:31:35,337 DEBUG [pool-7-thread-1] o.a.http.wire [Wire.java:72] http-outgoing-37 << "Date: Thu, 10 May 2018 13:31:35 GMT[\r][\n]"
2018-05-10 09:31:35,337 DEBUG [pool-7-thread-1] o.a.http.wire [Wire.java:72] http-outgoing-37 << "Content-Type: text/csv; charset=utf-8[\r][\n]"
2018-05-10 09:31:35,337 DEBUG [pool-7-thread-1] o.a.http.wire [Wire.java:72] http-outgoing-37 << "Content-Length: 15[\r][\n]"
2018-05-10 09:31:35,338 DEBUG [pool-7-thread-1] o.a.http.wire [Wire.java:72] http-outgoing-37 << "Connection: keep-alive[\r][\n]"
2018-05-10 09:31:35,338 DEBUG [pool-7-thread-1] o.a.http.wire [Wire.java:72] http-outgoing-37 << "x-amzn-RequestId: 1b2139c5-305c-41f0-b32e-556979c1177b[\r][\n]"
2018-05-10 09:31:35,338 DEBUG [pool-7-thread-1] o.a.http.wire [Wire.java:72] http-outgoing-37 << "x-Amzn-Invoked-Production-Variant: initialInstance[\r][\n]"
2018-05-10 09:31:35,338 DEBUG [pool-7-thread-1] o.a.http.wire [Wire.java:72] http-outgoing-37 << "[\r][\n]"
2018-05-10 09:31:35,338 DEBUG [pool-7-thread-1] o.a.http.wire [Wire.java:86] http-outgoing-37 << "0.0208577029407"
```

Since this isn't an umarshalling issue, I would suggest opening a ticket with Support so they can get your issue forwarded to the SageMaker team. The service team will be able to tell you whether this is expected behavior or not.

**Comment by lujea** [comment]
Thanks for the feedback Dongie.

**Comment by lujea** [comment]
@dagnir I tried the link you gave for the support (https://console.aws.amazon.com/support/), however it does not seem that I can create a ticket with the current plan that I have. Is it at least possible to post on the forum ?

**Comment by dagnir** [comment]
Ah sorry about that. Yes, a forum post should also help get your issue in front of the team.
